include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,337 @@
+            __running = __binary_op(__running, sycl::reduce_over_group(__subgroup, __contribution, __binary_op));
+
+            // If we found a full value, we can stop looking at previous tiles. Otherwise,
+            // keep going through tiles until we either find a full tile or we've completely
+            // recomputed the prefix using partial values
+            if (__is_full_ballot_bits)
+                break;
+        }
+        return __running;
+    }
+
+    _AtomicRefT __atomic_flag;
+};
+
+template <typename _KernelName>
+struct __lookback_init_submitter;
+
+template <typename... _Name>
+struct __lookback_init_submitter<oneapi::dpl::__par_backend_hetero::__internal::__optional_kernel_name<_Name...>>
+{
+    template <typename _StatusFlags, typename _Flag>
+    sycl::event
+    operator()(sycl::queue __q, _StatusFlags&& __status_flags, ::std::size_t __status_flags_size,
+               ::std::uint16_t __status_flag_padding, _Flag __oob_value) const
+    {
+        return __q.submit([&](sycl::handler& __hdl) {
+            __hdl.parallel_for<_Name...>(sycl::range<1>{__status_flags_size}, [=](const sycl::item<1>& __item) {
+                auto __id = __item.get_linear_id();
+                __status_flags[__id] = __id < __status_flag_padding ? __oob_value : 0;
+            });
+        });
+    }
+};
+
+template <::std::uint16_t __data_per_workitem, ::std::uint16_t __workgroup_size, typename _Type, typename _FlagType,
+          typename _KernelName>
+struct __lookback_submitter;
+
+template <::std::uint16_t __data_per_workitem, ::std::uint16_t __workgroup_size, typename _Type, typename _FlagType,
+          typename... _Name>
+struct __lookback_submitter<__data_per_workitem, __workgroup_size, _Type, _FlagType,
+                            oneapi::dpl::__par_backend_hetero::__internal::__optional_kernel_name<_Name...>>
+{
+    using _FlagStorageType = typename _FlagType::_StorageType;
+    static constexpr std::uint32_t __elems_in_tile = __workgroup_size * __data_per_workitem;
+
+    template <typename _InRng, typename _OutRng, typename _BinaryOp, typename _StatusFlags>
+    sycl::event
+    operator()(sycl::queue __q, sycl::event __prev_event, _InRng&& __in_rng, _OutRng&& __out_rng, _BinaryOp __binary_op,
+               ::std::size_t __n, _StatusFlags&& __status_flags, ::std::size_t __status_flags_size,
+               ::std::size_t __current_num_items) const
+    {
+        return __q.submit([&](sycl::handler& __hdl) {
+            auto __tile_vals = sycl::local_accessor<_Type, 1>(sycl::range<1>{__elems_in_tile}, __hdl);
+            __hdl.depends_on(__prev_event);
+
+            oneapi::dpl::__ranges::__require_access(__hdl, __in_rng, __out_rng);
+            __hdl.parallel_for<_Name...>(
+                sycl::nd_range<1>(__current_num_items, __workgroup_size),
+                [=](const sycl::nd_item<1>& __item) [[intel::reqd_sub_group_size(SUBGROUP_SIZE)]] {
```

Comment:
Usage of `[[_ONEDPL_SYCL_REQD_SUB_GROUP_SIZE_IF_SUPPORTED(__req_sub_group_size)]]` could potentially be problematic here since other portions of code assume the subgroup size is always `32`. If we want this solution to be device generic, I think we would also need to query the subgroup size at runtime in the kernel to be correct.